command/alloc_logs.go

```diff
@@ -195,77 +208,95 @@ func (l *AllocLogsCommand) Run(args []string) int {
 			}
 		} else {
-			task, err = lookupAllocTask(alloc)
+			l.task, err = lookupAllocTask(alloc)
 		}
 	}
 	if err != nil {
 		l.Ui.Error(fmt.Sprintf("Failed to validate task: %s", err))
 		return 1
 	}
 
-	logType := "stdout"
-	if stderr {
-		logType = "stderr"
+	// In order to run the mixed log output, we can only follow the files from
+	// their current positions. There is no way to interleave previous log
+	// lines as there is no timestamp references.
+	if l.follow && !(l.stderr || l.stdout || l.tail || l.numLines > 0 || l.numBytes > 0) {
+		if err := l.tailMultipleFiles(client, alloc); err != nil {
+			l.Ui.Error(fmt.Sprintf("Failed to tail stdout and stderr files: %v", err))
+			return 1
+		}
+	} else {
+
+		// If we are not strictly following the two files, we cannot support
+		// specifying both are targets.
+		if l.stderr && l.stdout {
+			l.Ui.Error("Unable to support both stdout and stderr")
+			return 1
+		}
+
+		logType := api.FSLogNameStdout
+		if l.stderr {
+			logType = api.FSLogNameStderr
+		}
+		if err := l.handleSingleFile(client, alloc, logType); err != nil {
+			l.Ui.Error(fmt.Sprintf("Failed to read %s file: %v", logType, err))
+			return 1
+		}
 	}
 
+	return 0
+}
+
+func (l *AllocLogsCommand) handleSingleFile(client *api.Client, alloc *api.Allocation, logType string) error {
 	// We have a file, output it.
 	var r io.ReadCloser
 	var readErr error
-	if !tail {
-		r, readErr = l.followFile(client, alloc, follow, task, logType, api.OriginStart, 0)
+	if !l.tail {
+		r, readErr = l.followFile(client, alloc, l.follow, l.task, logType, api.OriginStart, 0)
 		if readErr != nil {
-			readErr = fmt.Errorf("Error reading file: %v", readErr)
+			return fmt.Errorf("error reading file: %v", readErr)
 		}
 	} else {
 		// Parse the offset
-		var offset int64 = defaultTailLines * bytesToLines
+		var offset = defaultTailLines * bytesToLines
 
-		if nLines, nBytes := numLines != -1, numBytes != -1; nLines && nBytes {
-			l.Ui.Error("Both -n and -c set")
-			return 1
+		if nLines, nBytes := l.numLines != -1, l.numBytes != -1; nLines && nBytes {
+			return errors.New("both -n and -c set")
 		} else if nLines {
-			offset = numLines * bytesToLines
+			offset = l.numLines * bytesToLines
 		} else if nBytes {
-			offset = numBytes
+			offset = l.numBytes
 		} else {
-			numLines = defaultTailLines
+			l.numLines = defaultTailLines
 		}
 
-		r, readErr = l.followFile(client, alloc, follow, task, logType, api.OriginEnd, offset)
+		r, readErr = l.followFile(client, alloc, l.follow, l.task, logType, api.OriginEnd, offset)
 
 		// If numLines is set, wrap the reader
-		if numLines != -1 {
-			r = NewLineLimitReader(r, int(numLines), int(numLines*bytesToLines), 1*time.Second)
+		if l.numLines != -1 {
+			r = NewLineLimitReader(r, int(l.numLines), int(l.numLines*bytesToLines), 1*time.Second)
 		}
 
 		if readErr != nil {
-			readErr = fmt.Errorf("Error tailing file: %v", readErr)
+			return fmt.Errorf("error tailing file: %v", readErr)
 		}
 	}
```

Comment:
Not part of the PR, but it may be worth flipping the lines to check for `readErr != nil` earlier since `r` will be `nil`